borg/repository.py

```diff
@@ -84,6 +84,11 @@ def __enter__(self):
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         if exc_type is not None:
+            no_space_left_on_device = exc_type is OSError and exc_val.errno == errno.ENOSPC
+            # The ENOSPC could have originated somewhere else besides the Repository. The cleanup is always safe, unless
+            # EIO or FS corruption ensues, which is why we specifically check for ENOSPC.
+            if self._active_txn and no_space_left_on_device:
+                self.io.cleanup(self.io.get_segments_transaction_id())
             self.rollback()
```

Comment:
i think we should log an error here also, like "no space left. rolling back uncommitted segments to free space."